File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(!head || !k) return head;

        ListNode* tail = head;
        int len = 1;

        while(tail->next){
            len++;
            tail = tail->next;
        }

        if(k % len == 0) return head;
        k = k % len;

        // make circular
        tail->next = head;

        ListNode* pivot = tail;
        int cnt = len - k;

        while(cnt--){
            pivot = pivot->next;
        }

        head = pivot->next;
        pivot->next = nullptr;

        return head;
    }
};

File: 0061-rotate-list/0061-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next:
            return head  
        l = 0
        temp = head
        while temp:
            l += 1
            temp = temp.next
        k = k % l
        if k == 0:
            return head
        t = l - k
        traverse = head
        for x in range(t - 1):
            traverse = traverse.next
        ans = traverse.next
        traverse.next = None   
        dummy = ans
        while dummy.next:
            dummy = dummy.next 
        dummy.next = head
        return ans
         

